crates/matrix-sdk/src/event_cache/mod.rs

```diff
@@ -283,23 +271,16 @@ impl EventCacheInner {
     /// Return a room-specific view over the [`EventCache`].
     ///
     /// It may not be found, if the room isn't known to the client.
-    async fn for_room(&self, room_id: &RoomId) -> Result<RoomEventCache> {
-        let by_room_guard = self.by_room.read().await;
-
-        match by_room_guard.get(room_id) {
-            Some(room) => Ok(room.clone()),
-            None => {
-                let room_event_cache = RoomEventCache::new(room_id.to_owned(), self.store.clone());
-
-                // Drop the `RwLockReadGuard`.
-                drop(by_room_guard);
-
+    async fn for_room(&self, room_id: &RoomId) -> RoomEventCache {
+        self.by_room
```

Comment:
Would it make sense to repeat the check that a room exists once the write lock was acquired instead of taking a `write()` lock always?

Reply:
[`BTreeMap::entry`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.entry) needs a `&mut self`.

Reply:
I mean, use the `get()` first, then if that fails, take the write lock and use `entry()` to check if an entry exists again. This would avoid the problem you are fixing without always taking a `write()` lock.

Reply:
Oh OK, I get it. It's 2 accesses though. Is it cheaper than acquiring a write lock? Hmm, likely yes.

Reply:
Done.